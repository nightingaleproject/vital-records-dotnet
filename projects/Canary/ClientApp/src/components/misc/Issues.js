import React, { Component } from 'react';
import { Icon, Message, Transition } from 'semantic-ui-react';

export class Issues extends Component {
  displayName = Issues.name;

  render() {
    return (
      <React.Fragment>
        {!!this.props.issues && this.props.issues.length > 0 && (
          <div className="inherit-width p-b-50">
            {this.props.issues.map(function(issue, index) {
              return (
                <Transition key={`issue-t-${index}`} transitionOnMount animation="fade" duration={1000}>
                  <div className="inherit-width p-b-10">
                    <Message icon size="large" negative={issue.severity.toLowerCase() === 'error'} warning={issue.severity.toLowerCase() === 'warning'}>
                      <Icon name="exclamation triangle" />
                      <Message.Content>{`${issue.message}`}</Message.Content>
                    </Message>
                  </div>
                </Transition>
              );
            })}
          </div>
        )}
      </React.Fragment>
    );
  }
}
