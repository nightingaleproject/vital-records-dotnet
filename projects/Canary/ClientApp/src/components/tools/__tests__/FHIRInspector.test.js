import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import { unmountComponentAtNode } from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import { FHIRInspector } from '../FHIRInspector';
import { Getter } from '../../misc/Getter';

const getterSpy = jest.spyOn(Getter.prototype, 'submitPaste');
let container = null;

beforeAll(() => {
  window.API_URL = '';
});

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders', () => {
  const tree = renderer.create(
    <MemoryRouter>
      <FHIRInspector recordType={"bfdr-birth"} recordTypeReadable={"BFDR Birth"} />
    </MemoryRouter>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should show the success message when issues is defined but empty', async () => {
  getterSpy.mockImplementationOnce(() => {
    const self = getterSpy.mock.instances[0];
    self.props.updateRecord({ fhirInfo: {} }, []);
  });
  const component = render(
    <MemoryRouter>
      <FHIRInspector recordType={"bfdr-birth"} recordTypeReadable={"BFDR Birth"} />
    </MemoryRouter>
    , container);
  const user = userEvent.setup();
  const submitButton = await component.findByText('Submit');
  const textArea = await component.findByRole('textbox');
  textArea.focus();
  await userEvent.paste('{ "good": true }');
  fireEvent.click(submitButton);
  const successMessage = await component.findByText('No issues were found!');
  expect(successMessage).toBeDefined();
});

it('should show the problems when there are issues', async () => {
  getterSpy.mockImplementationOnce(() => {
    const self = getterSpy.mock.instances[0];
    self.props.updateRecord('',
      [{ severity: 'error', message: 'Problem description: data no good.' },
      { severity: 'warning', message: 'Also, I am tired.' }]);
  });
  const component = render(
    <MemoryRouter>
      <FHIRInspector recordType={"bfdr-birth"} recordTypeReadable={"BFDR Birth"} />
    </MemoryRouter>, container);
  const user = userEvent.setup();
  const submitButton = await component.findByText('Submit');
  const textArea = await component.findByRole('textbox');
  textArea.focus();
  userEvent.paste('{ "good": false }').then(() => fireEvent.click(submitButton)); // logs error from act?
  const errorMessage = await component.findByText('Problem description: data no good.');
  expect(errorMessage).toBeDefined();
  const warningMessage = await component.findByText('Also, I am tired.');
  expect(warningMessage).toBeDefined();
  await expect(component.findByText('No issues were found!')).rejects.toBeDefined();
});

it('should show issues when record is truthy and there is at least one issue', async () => {
  getterSpy.mockImplementationOnce(() => {
    const self = getterSpy.mock.instances[0];
    self.props.updateRecord({
      fhirInfo: {
        "Child Demographics": {}
      }
    },
      [{ severity: 'error', message: 'Problem description: data no good.' },
      { severity: 'warning', message: 'Also, I am tired.' }]);
  });
  const component = render(
    <MemoryRouter>
      <FHIRInspector recordType={"bfdr-birth"} recordTypeReadable={"BFDR Birth"} />
    </MemoryRouter>, container);
  const user = userEvent.setup();
  const submitButton = await component.findByText('Submit');
  const textArea = await component.findByRole('textbox');
  textArea.focus();
  userEvent.paste('{ "good": false }').then(() => fireEvent.click(submitButton));
  const errorMessage = await component.findByText('Problem description: data no good.');
  expect(errorMessage).toBeDefined();
  const warningMessage = await component.findByText('Also, I am tired.');
  expect(warningMessage).toBeDefined();
  const recordSection = await component.findByText('Child Demographics');
  expect(recordSection).toBeDefined();
});