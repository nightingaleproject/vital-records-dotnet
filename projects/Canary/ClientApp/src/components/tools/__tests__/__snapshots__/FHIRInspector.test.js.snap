// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="ui grid"
>
  <div
    className="row"
  >
    <div
      className="ui breadcrumb"
    >
      <a
        className="section"
        href="/bfdr-birth"
        onClick={[Function]}
      >
        Dashboard
      </a>
      <i
        aria-hidden="true"
        className="right chevron icon divider"
        onClick={[Function]}
      />
      <div
        className="section"
        onClick={[Function]}
      >
        FHIR 
        BFDR Birth
         Record Inspector
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <form
      className="ui form p-t-10"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <div
          className="ui buttons"
        >
          <button
            className="ui primary button"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="clipboard icon"
              onClick={[Function]}
            />
            Copy & Paste
          </button>
          <div
            className="or"
          />
          <button
            className="ui button"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="upload icon"
              onClick={[Function]}
            />
            Upload
          </button>
          <div
            className="or"
          />
          <button
            className="ui button"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="send icon"
              onClick={[Function]}
            />
            POST to Canary
          </button>
        </div>
        <div
          className="ui left pointing basic label m-l-15"
          onClick={[Function]}
        >
          How would you like to import the record into Canary?
        </div>
      </div>
    </form>
    <div
      className="ui segment"
    >
      <div
        className="ui inverted dimmer"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="ui big text loader"
          >
            Processing
          </div>
        </div>
      </div>
      <div
        className="ui container"
      >
        <form
          className="ui form"
          onSubmit={[Function]}
        >
          <h4>
            Paste the 
            record
             contents below. 
            The contents can be formatted as FHIR XML or FHIR JSON.
          </h4>
          <div
            id="paste-input"
            style={
              Object {
                "height": "400px",
                "width": "100%",
              }
            }
          />
          <button
            className="ui primary left floated button m-t-10"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="code icon"
              onClick={[Function]}
            />
            Format Input
          </button>
          <button
            className="ui primary disabled right floated button m-t-10"
            disabled={true}
            onClick={[Function]}
            tabIndex={-1}
          >
            <i
              aria-hidden="true"
              className="send icon"
              onClick={[Function]}
            />
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
  <div
    className="p-b-15"
  />
</div>
`;
